File: linkedlist.c
#include <stdio.h>
#include <stdlib.h>

struct linkedList {
    int label;
    struct linkedList *next;
};

struct linkedList *lookup(struct linkedList *list, int label) {
    for (; list != NULL && list->label != label; list = list->next)
        ;
    return list;
}

struct linkedList *add(struct linkedList *list, int label) {
    struct linkedList *node = NULL;
    node = (struct linkedList *) malloc(sizeof(*node));
    node->label = label;
    node->next = list;
    return node;
}

struct linkedList *append(struct linkedList *list, int label) {
    if (list == NULL) {
        list = add(list, label);
    } else if (list->label != label) {
        struct linkedList *node = list;
        for (; node->next != NULL; node = node->next)
            ;
        if (node->next == NULL) {
            node->next = add(node->next, label);
        }
    }
    return list;
}

void deleteNode(struct linkedList *list, int label) {
    struct linkedList *temp;
    if (list->label == label) {
        temp = list->next;
        *list = *temp;
        free((void *) temp);
        return;
    }
    for (; list->next != NULL; list = list->next) {
        if (list->next->label == label) {
            temp = list->next;
            list->next = list->next->next;
            free((void *) temp);
        }
    }
}

struct linkedList *reverse(struct linkedList *head) {
    struct linkedList *prev, *cur, *next;
    prev = NULL;
    cur = head;
    while (cur != NULL) {
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }
    return prev;
}

void inplace_reverse(struct linkedList **head) {
    struct linkedList *prev, *cur, *next;
    prev = NULL;
    cur = *head;
    while (cur != NULL) {
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }
    *head = prev;
}

void show(struct linkedList *head) {
    for (; head != NULL; head = head->next) {
        printf("%d->", head->label);
    }
    printf("NULL\n");
}

int main() {
    int i;
    struct linkedList *foo = NULL, *bar = NULL;
    for (i = 1; i < 10; i++) {
        foo = add(foo, i);
        bar = append(bar, i);
    }
    printf("foo: ");
    show(foo);
    printf("bar: ");
    show(bar);
    inplace_reverse(&bar);
    printf("inplace reverse bar: ");
    show(bar);
    bar = reverse(bar);
    printf("reverse bar: ");
    show(bar);
    deleteNode(foo, 9);
    printf("delete head 9 from foo: ");
    show(foo);
    deleteNode(foo, 5);
    printf("delete 5 from foo: ");
    show(foo);
    return 0;
}

File: basic_data_structure/linkedlist.py

class node:
    def __init__(self, label, next=None):
        self.label = label
        self.next = next
    def __str__(self):
        return str(self.label)

class linkedlist:
    def __init__(self):
        self.head = None

    def add(self, label):
        self.head = node(label, self.head)

    def is_empty(self):
        return self.head == None

    def append(self, label):
        head = self.head
        if not head:
            self.add(label)
        else:
            while head.next != None:
                head = head.next
            head.next = node(label)

    def reverse(self):
        prev = None
        cur = self.head
        while cur:
            next = cur.next 
            cur.next = prev
            prev = cur
            cur = next
        self.head = prev
    
    def __add__(self, other):
        if self.head is None:
            self.head = other.head 
            return
        tail = self.head
        while tail.next:
            tail = tail.next 
        tail.next = other.head

    def lookup(self, label):
        head = self.head
        while head != None and head.label != label:
            head = head.next
        return head

    def delete(self, label):
        if self.head.label == label:
            self.head = self.head.next
        else:
            head = self.head
            while head.next != None:
                if head.next.label == label:
                   head.next = head.next.next 
                   break
                head = head.next

    def __eq__(self, other):
        cur1 = self.head
        cur2 = other.head
        while cur1 and cur2: 
            if cur1.label != cur2.label:
                return False
            cur1 = cur1.next
            cur2 = cur2.next
        return cur1 is None and cur2 is None

    def __str__(self):
        head = self.head
        s = ''
        while head != None:
            s += '%s->'%str(head.label) 
            head = head.next
        s += 'None'
        return s
        
if __name__ == '__main__':
    foo = linkedlist()
    bar = linkedlist()
    for i in range(1, 10):
        foo.append(i)
        bar.add(i)
    print('foo: %s'%str(foo))
    print('bar: %s'%str(bar))
    print('foo == bar ? ', end='')
    print(foo == bar)
    foo.reverse()
    print('reverse foo: %s'%str(foo))
    print('foo == bar ? ', end='')
    print(foo == bar)

    head = foo.lookup(6)
    print('6 in foo: %s'%str(head))

    foo.delete(4)
    print('foo delete 4: %s'%str(foo))
